Zend/zend_gc.c

```diff
@@ -829,6 +866,33 @@ static void gc_mark_grey(zend_refcounted *ref, gc_stack *stack)
 			zval *table;
 			int len;
 
+			if (UNEXPECTED(GC_FLAGS(obj) & IS_OBJ_WEAKLY_REFERENCED)) {
+				zend_weakmap_get_object_entries_gc(obj, &table, &len);
+				n = len;
+				zv = table;
+				for (; n != 0; n-=2) {
+					ZEND_ASSERT(Z_TYPE_P(zv) == IS_PTR);
+					zval *value = (zval*) Z_PTR_P(zv);
+					zval *weakmap = zv+1;
+					ZEND_ASSERT(Z_REFCOUNTED_P(weakmap));
+					/* Skip reference if already followed through weakmap */
+					if (!GC_REF_CHECK_COLOR(Z_COUNTED_P(weakmap), GC_GREY)) {
+						if (Z_REFCOUNTED_P(value)) {
+							ZEND_ASSERT(!GC_FROM_WEAKMAP_KEY(value));
+							GC_SET_FROM_WEAKMAP_KEY(value);
```

Comment:
In the first phase, we skip virtual references if the WeakMap is already scanned. Otherwise, we mark the entry with GC_SET_FROM_WEAKMAP_KEY. This sets which virtual references can be followed in the later phases.